File: python/ecuaciones_no_lineales/codigos/newton-raphson.py
from sympy import *
import matplotlib.pyplot


def newton_raphson(f, xo, tol, iterMax):
    f1 = sympify(f)
    df = diff(f1)
    error = tol + 1
    k = 0
    it = []
    e = []
    while (error > tol and k < iterMax):
        k = k + 1
        deno = df.subs('x', xo)
        if (deno == 0):
            x = []
            k = []
            error = []
            print('La funcion se indefine.')
            break
        else:
            nume = f1.subs('x', xo)
            xn = xo - nume/deno
            error = abs(xn)
            it.append(k)
            e.append(error)
            xo = xn
    matplotlib.pyplot.plot(it, e)
    ans = {
        "x": xn,
        "k": k,
        "error": error
    }
    return ans

# newton_raphson('cos(2*x)**2-x**2',3/4,10**-8,10)

File: octave/ecuaciones_no_lineales/newton_raphson.m
function [xn, k, error] = newton_raphson(f, xo, tol, iterMax)
    syms x;
    f1 = str2func(f); % 1
    % f1 = matlabFunction(sym(f)); % 2

    df = matlabFunction(diff(sym(f1)));
    k = 0;
    error = tol + 1;
    e = [];

    while (tol < error && k < iterMax)
        deno = df(xo);
        
        if (deno == 0)
            x = [];
            k = []; 
            error = [];
            disp('La función se indefine.')
            break;
        else
            xn = xo - (f1(xo) / deno);
            error = abs(xo - (f1(xo) / deno));
            e = [e error];
            xo = xn;
            ++k;
        end
    end

    plot(1 : k, e)
    xlabel('iter (k)')
    ylabel('Error (|f(x_k)|)')
    title('Error del Metodo de la Newton-Raphson')
end

% 1
% [x, k, error] = newton_raphson('@(x)cos(2*x).^2-x.^2', 3/4, 10^-8, 10)

% 2
% [x, k, error] = newton_raphson(cos(2*x).^2-x.^2, 3/4, 10^-8, 10)